engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java

```diff
@@ -0,0 +1,163 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.node.expression.builtin.error.PolyglotError;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "immutable", stdlibName = "Standard.Base.Data.Vector.Vector")
+public final class Vector implements TruffleObject {
+  private final Object storage;
+
+  private Vector(Object storage) {
+    assert InteropLibrary.getUncached().hasArrayElements(storage)
+        : "Vector needs array-like delegate, but got: " + storage;
+    this.storage = storage;
+  }
+
+  @Builtin.Method(
+      name = "new_builtin",
+      description = "Creates new Vector with given length and provided elements.")
+  @Builtin.Specialize
+  public static Object newFromFunction(long length, Function fun, InteropLibrary interop) {
+    Object[] target = new Object[(int) length];
+    for (int i = 0; i < length; i++) {
```

Comment:
This seems dangerous to me.

We are using `int` for `i`, but comparing with `long length`. In case `length` is so big that it cannot fit into `int`, I'm pretty sure this will become an infinite loop (because `i++` will start overflowing and will never be able to exceed `length`).

I'd suggest a safe cast of `length` to `int` and throwing an exception that the array is too large to allocate.

Reply:
Using `Math.toIntExact` is good idea: a6d9d306312f